content/blog/0021-bundle-accounting.mdx

```diff
@@ -0,0 +1,137 @@
+  );
+  return {
+    pending: queries.some((_) => _.isPending),
+    initialRows,
+    initialSelected: Array.filterMap(initialRows, (_) =>
+      isRecipient(_) ? Option.some(_.id) : Option.none(),
+    ),
+  };
+};
+```
+
+This function takes the result of two collections that have some possible overlap, cleans up `users` from members that don't have an email, concatenates it to the `recipients` into a nice wholesome `(User | Recipient)[]` that is unusable, uses an `@effect/schema` `NormalizedRow` schema to homogenize it into a useful `RecipientRow[]`, and remove duplicate rows, keeping only the `User`-originated ones.
+
+This utility also then creates a derived collection that is just the ids of the `RecipientRows` that either represent a `Recipient` *or* represent a `User` *that has a corresponding `Recipient`* (remember we deduped them).
+
+You could glue a solution that does this using `lodash`, `zod`, and some hand-crafted stuff, but it won't be this elegant considering the definition above. Especially the normalization part and the types. In this snippet it looks like it's trivial -- it's not. This function was over 100 LOC before.
+
+Recently, I have seen `effect` being referred to as "a different language" or "hard to learn" or "unreadable", including commentary from people who have never even used the library. Effect is a complete toolkit for developing enterprise-grade applications. Not an RxJS simulator or a cool way to use `Result` from Rust.
+
+I also have a ton of `Effect`-effect usage. With all the bells and whistles! From a quick search here's a partial list of modules I use in this project
+
+- `Effect`
+- `Layer`
+- `Context`
+- `Predicate`
+- `Array`
+- `Cause`
+- `Option`
+- `Config`
+- `Scope`
+- `ManagedRuntime`
+- `GlobalValue`
+- `String`
+- `HttpClient`
+- `ClientRequest`
+- `ClientResponse`
+- `Schema`
+- `Data`
+
+"Oh no! Egad! Thats too much to learn!" You're saying right now, totally ignoring that you probably have learned all of these concepts, and much more, piecemeal, from scratch on every project according to whatever 30-or-so packages are included in the project.
+
+Don't worry dear reader, you don't have to learn any of these, you can be productive
+with `effect` even when using just a single module, and learn to use the rest at your own pace. For me its the missing standard library that is internally consistent and 100% interoperable.
+
+Cut to yesterday, I've finally finished implementing all the features I needed. 80 changed files (mostly new files), around 12K LOC added by yours truly (trimmed a lot of fat thanks to `effect`, but react theres still a lot of React code and CSS).
+
+So, nows the money time. How much did I pay for my hubris?
+
+[net 50kb difference](/public/images/datner-bundle-50kb.png)
+
+Net 50kb.
+
+`effect`, `@effect/schema`, `@effect/platform`, `@effect/platform-browser`, and `react-query` to boot.
+
+80 files, 4 new routes, nearly all new code.
+
```

Comment:
```suggestion

As `effect` makes its way into more and more areas of your application, its size amortizes because you inevitably trim a lot of "fat" from your application. I removed almost all the custom utilities I had previously built and was able to uninstall several packages.

Additionally, `effect` is 100% tree-shakeable, so I didn't end up with unnecessary code in my final application bundle.

I can say without hesitation that the `effect` "tax" was 100% worth it.
```

Reply:
a bit too plastic-y, don't you think?

Reply:
Edit as you see fit :) this was just a suggestion.